propolis/src/block/file.rs

```diff
@@ -0,0 +1,228 @@
+        }
+    }
+
+    async fn do_scheduling(&self, actx: &mut AsyncCtx) {
+        loop {
+            let avail = self.idle_threads.acquire().await.unwrap();
+            avail.forget();
+
+            if let Some(req) = self.waiter.next(self.dev.as_ref(), actx).await {
+                let mut queue = self.queue.lock().unwrap();
+                queue.push_back(req);
+                drop(queue);
+                self.cv.notify_one();
+            }
+        }
+    }
+
+    fn spawn(self: &Arc<Self>, worker_count: NonZeroUsize, disp: &Dispatcher) {
+        for i in 0..worker_count.get() {
+            let tself = Arc::clone(self);
+
+            // Configure a waker to help threads to reach their yield points
+            // Doing this once (from thread 0) is adequate to wake them all.
+            let wake = if i == 0 {
+                let tnotify = Arc::downgrade(self);
+                Some(Box::new(move |_ctx: &DispCtx| {
+                    if let Some(this) = tnotify.upgrade() {
+                        let _guard = this.queue.lock().unwrap();
+                        this.cv.notify_all();
+                    }
+                }) as Box<WakeFn>)
+            } else {
+                None
+            };
+
+            let _ = disp
+                .spawn_sync(
+                    format!("file bdev worker {}", i),
+                    Box::new(move |mut sctx| {
+                        tself.blocking_loop(&mut sctx);
+                    }),
+                    wake,
+                )
+                .unwrap();
+        }
+
+        // TODO: do we need the task for later?
+        let sched_self = Arc::clone(self);
+        let _sched_task = disp.spawn_async(|mut actx| async move {
+            let _ = sched_self.do_scheduling(&mut actx).await;
+        });
+    }
+}
+
+fn process_request(
+    fd: RawFd,
+    req: &block::Request,
+    ctx: &DispCtx,
+) -> Result<()> {
+    let mem = ctx.mctx.memctx();
+    match req.oper() {
+        block::Operation::Read(off) => {
+            let maps = req.mappings(&mem).ok_or_else(|| {
+                Error::new(ErrorKind::Other, "bad guest region")
+            })?;
+
+            let nbytes = maps.preadv(fd, off as i64)?;
+            if nbytes != req.len() {
+                return Err(Error::new(ErrorKind::Other, "bad read length"));
+            }
+        }
+        block::Operation::Write(off) => {
+            let maps = req.mappings(&mem).ok_or_else(|| {
+                Error::new(ErrorKind::Other, "bad guest region")
+            })?;
+
+            let nbytes = maps.pwritev(fd, off as i64)?;
+            if nbytes != req.len() {
+                return Err(Error::new(ErrorKind::Other, "bad write length"));
+            }
+        }
+        block::Operation::Flush(_off, _len) => {
+            // SAFETY: The backing fd should be valid and fdatasync() is
+            // doing nothing besides the flush.
+            unsafe {
+                let res = fdatasync(fd);
```

Comment:
should use `fp.sync_data` if possible: https://github.com/oxidecomputer/propolis/pull/44

Reply:
Done